### DISCUSSION TITLE
can't get sensor to run - need your help please

### DISCUSSION THREAD

**Discussion by steuerzentrale** [PricelessToolkit/MailBoxGuard#26] [Q&A] [answered]
Hi there

First of all many thanks for this nice piece of hardware, the build-quality is nothing but excellent! I want to use the sensor with my MQTT-broker (Mosquito) on my Home Assistant instance. Unfortunately I struggle to connect the sensor and the Lora Gateway. I could upload the corresponding sketches onto the devices and MQTT sees the gateway. I live in Switzerland where the frequency is 868Mhz.

Lora-Gateway: LilyGo TTGO LoRa32 868Mhz Version T3_V1.6.1
mailbox-sensor: powered via **USB-C** for testing purposes

Arduino-sketch Mailbox_Guard_Sensor_MQTT_Auto_Discovery.ino. **I just post the sections I have modified in the code.**

`#define SignalBandwidth 125E3`
`#define SpreadingFactor 12`
`#define TransmitBattPercent 1`
`#define CodingRate 8`
`#define SyncWord 0xF3`
`#define PreambleLength 8`
`#define TxPower 20`
`#define BAND 868E6     // frequency in Hz (ASIA 433E6, EU 868E6, US 915E6)`
`String NewMailCode = "0xA2B2"; // For Example "0xA2B2";`
`String LowBatteryCode = "0xLBAT"; // For Example "0xLBAT";`

  `LoRa.setSignalBandwidth(125E3);         // signal bandwidth in Hz, defaults to 125E3`
  `LoRa.setSpreadingFactor(12);                 // ranges from 6-12,default 7 see API docs`
`  LoRa.setCodingRate4(8);        // Supported values are between 5 and 8, these correspond to coding rates of 4/5 and 4/8. The coding rate numerator is fixed at 4.`
`  LoRa.setSyncWord(0xF3);                     // byte value to use as the sync word, defaults to 0x12`
`  LoRa.setPreambleLength(8);       //Supported values are between 6 and 65535.`
`  LoRa.disableCrc();                          // Enable or disable CRC usage, by default a CRC is not used LoRa.disableCrc();`
`  LoRa.setTxPower(20);                // TX power in dB, defaults to 17, Supported values are 2 to 20`

Arduino-sketch LoRa_Gateway_MQTT_Auto_Discovery.ino. **I just post the sections I have modified in the code.**

`#define SignalBandwidth 125E3`
`#define SpreadingFactor 12`
`#define CodingRate 8`
`#define SyncWord 0xF3`
`#define PreambleLength 8`
`#define TxPower 20`
`float BAND = 868E6; // 433E6 / 868E6 / 915E6 /`

`const char *ssid = "XX";`
`const char *password = "XX";`
`const char *mqtt_username = "XX";`
`const char *mqtt_password = "XX";`
`const char *mqtt_server = "192.168.1.116";`
`const int mqtt_port = 1883;`
`bool retain = true;`

`String NewMailCode = "0xA2B2"; // For Example "0xA2B2";`
`String LowBatteryCode = "0xLBAT"; // For Example "0xLBAT";`

`  LoRa.setSignalBandwidth(125E3); // signal bandwidth in Hz, defaults to 125E3`
`  LoRa.setSpreadingFactor(12); // ranges from 6-12,default 7 see API docs`
`  LoRa.setCodingRate4(8);          // Supported values are between 5 and 8, these correspond to coding rates of 4/5 and 4/8. The coding rate numerator is fixed at 4.`
`  LoRa.setSyncWord(0xF3);               // byte value to use as the sync word, defaults to 0x12`
`  LoRa.setPreambleLength(8);   // Supported values are between 6 and 65535.`
`  LoRa.disableCrc();                        // Enable or disable CRC usage, by default a CRC is not used LoRa.disableCrc();`
`  LoRa.setTxPower(20);                 // TX power in dB, defaults to 17, Supported values are 2 to 20`

extract from MQTT-explorer
<img width="565" alt="Bildschirmfoto 2023-11-19 um 12 16 45" src="https://github.com/PricelessToolkit/MailBoxGuard/assets/75168550/0703bc8c-7506-4edf-9ac9-d20e25ee588b">

I would appreciate your support, I cant figure out what I am doing wrong. I expected that when the PIR-sensor detects motion the state of the mailbox-entity in HA would change, but nothing happens.

all the best

Andrew

**Comment by PricelessToolkit** [comment]
Hi Andrew, Thank you for supporting my work!!

1. USB-C is not meant for powering the board, if you do not have a battery yet, connect external 3.3-4.2 v directly to battery contacts.
2.  In `board.h` enable v1.6.
3. 
## In the gateway sketch, only this part needs to be changed

```
///////////////////////////////////////////////////// CHANGE THIS /////////////////////////////////////////////////////////////

#define SignalBandwidth 125E3
#define SpreadingFactor 12
#define CodingRate 8
#define SyncWord 0xF3
#define PreambleLength 8
#define TxPower 20
float BAND = 868E6; // 433E6 / 868E6 / 915E6 /

const char *ssid = "Your_WIFI_SSID";
const char *password = "Your_WIFI_password";
const char *mqtt_username = "Your_mqtt_username";
const char *mqtt_password = "Your_mqtt_password";
const char *mqtt_server = "Your_mqtt/homeassistant server IP";
const int mqtt_port = 1883;
bool retain = true;

String NewMailCode = "REPLACE_WITH_NEW_MAIL_CODE"; // For Example "0xA2B2";
String LowBatteryCode = "REPLACE_WITH_LOW_BATTERY_CODE"; // For Example "0xLBAT";

// IMPORTANT: Set TransmitBattPercent to 1 in the Sensor Config!

String bat_val = "";
String received_code = "";

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```




## You modified the wrong part

 ```
 LoRa.setSignalBandwidth(125E3); // signal bandwidth in Hz, defaults to 125E3
  LoRa.setSpreadingFactor(12); // ranges from 6-12,default 7 see API docs
  LoRa.setCodingRate4(8);          // Supported values are between 5 and 8, these correspond to coding rates of 4/5 and 4/8. The coding rate numerator is fixed at 4.
  LoRa.setSyncWord(0xF3);               // byte value to use as the sync word, defaults to 0x12
  LoRa.setPreambleLength(8);   // Supported values are between 6 and 65535.
  LoRa.disableCrc();                        // Enable or disable CRC usage, by default a CRC is not used LoRa.disableCrc();
  LoRa.setTxPower(20);                 // TX power in dB, defaults to 17, Supported values are 2 to 20
```

## Original "Don't touch"

```
  LoRa.setSignalBandwidth(SignalBandwidth); // signal bandwidth in Hz, defaults to 125E3
  LoRa.setSpreadingFactor(SpreadingFactor); // ranges from 6-12,default 7 see API docs
  LoRa.setCodingRate4(CodingRate);          // Supported values are between 5 and 8, these correspond to coding rates of 4/5 and 4/8. The coding rate numerator is fixed at 4.
  LoRa.setSyncWord(SyncWord);               // byte value to use as the sync word, defaults to 0x12
  LoRa.setPreambleLength(PreambleLength);   // Supported values are between 6 and 65535.
  LoRa.disableCrc();                        // Enable or disable CRC usage, by default a CRC is not used LoRa.disableCrc();
  LoRa.setTxPower(TxPower);                 // TX power in dB, defaults to 17, Supported values are 2 to 20
```

## In the sensor sketch the same thing

**Comment by steuerzentrale** [comment]
many thanks for your quick reply. I just changed the above mentioned code on both devices and enabled **V1.6** in board.h (#define LORA_V1_6_OLED  1). I power the sensor via your UPDI-programmer from 3.3v to battery + and GND to battery -. I can see that the blue LED lights up when the PIR-sensor detects motion, so it seems that the sensor is basically working.

However I suspect the issue with MQTT which is odd because other devices via MQTT work fine. The gateway can reach my MQTT-broker (official HA add-on) but I have several connects and disconnects. there are no logs under settings/system/logs in HA. A restart of the add-on and HA-instance didnt help.

As other devices work fine via MQTT the issue comes probably from the gateway. The gateway gets a unique fixed IP-address.

**question**: Are there any parameters I could change in the gateway-sketch to address the issue? I found this thread which is more or less the same issue I have https://community.home-assistant.io/t/mqtt-with-esp32-gives-timeout-disconnecting/260298 unfortunately there is no solution provided.

**environment**
Mosquitto-broker 6.3.1
HA Core 2023.11.2
Supervisor 2023.11.3
Operating System 11.1
Frontend 20231030.2

<img width="1010" alt="Bildschirmfoto 2023-11-19 um 18 08 30" src="https://github.com/PricelessToolkit/MailBoxGuard/assets/75168550/a8948b3f-c255-409e-8ad8-80f16852d785">

**Reply by PricelessToolkit** [reply]
1. I think disconnecting from MQTT is normal because after publishing "client.end Publish();" is executed. 
2. Please plug the gateway via USB and open the serial monitor, you will see LoRa data when it receives something and statuses MQTT Connected... Error...
3.  If you connected UNIProg 3.3v to a 3.3v sensor, it will ONLY trigger! 1 time, then you will need to disconnect and reconnect UNIProg USB.

**Reply by PricelessToolkit** [reply]
Please zip modified sketches for the sensor and for the gateway, and upload them here [https://share.iot-dev.me/g/Eb8DWDuHiqoXz4gW](https://share.iot-dev.me/g/Eb8DWDuHiqoXz4gW)  I will test it tomorrow.

**Comment by steuerzentrale** [comment]
thats great thanks a lot! I will upload it ASAP really appreciate your support!

here you go, gateway https://share.iot-dev.me/-sDj7A84Vnh / mailbox-sensor https://share.iot-dev.me/-x5VKy5gkiL

**Reply by PricelessToolkit** [reply]
Hi please can  you test this part
( Plug the gateway via USB and open the serial monitor, you will see LoRa data when it receives something and statuses MQTT Connected... Error... )

Can you see the NewMailCode??

**Comment by steuerzentrale** [comment, marked as answer]
that was a good advice with the serial monitor, thanks. I uploaded the output (word-file) here:  https://1drv.ms/w/s!Ald4PcI_y0i6h64mh7IKfBiX2i53NQ?e=iQQPWk 

I cant see the NewMailCode despite the mailbox-sensor is powered and the PIR-sensor detects motion. What I can see that the gateway speaks to the MQTT-server. could you read anything helpful from the serial output log?
 
**UPDATE:** I finally could get it running. I uninstalled Arduino IDE and did a fresh install (legacy version 1.8). Then I pasted your libraries in the Arduino-folder and installed the two boards (ESP32 Version 2.0.14 and megaTiny Core). I put in my information/credential and uploded the sketch to the gateway.

two final questions

Q1: Is it intended that the binary_sensor.mailbox always displays open in HA even when there is no motion detected?
<img width="389" alt="Bildschirmfoto 2023-11-21 um 10 10 18" src="https://github.com/PricelessToolkit/MailBoxGuard/assets/75168550/1a5c72a3-0d32-4363-86dc-725b0d716527">

Q2: Can I somehow get the state of the newmail-code (0xA2B2) in an entity within HA? You suggest to create an entity which reads LoRa-Gateway/Code but I use Auto-MQTT and dont have those entities.
<img width="902" alt="Bildschirmfoto 2023-11-21 um 11 32 15" src="https://github.com/PricelessToolkit/MailBoxGuard/assets/75168550/f75a0517-544d-4c9d-b62e-53cb3a72a53b">



![Screenshot 2023-11-21 095625](https://github.com/PricelessToolkit/MailBoxGuard/assets/75168550/08d450ea-088f-4347-b7ac-3bb3f0164ab4)

**Reply by PricelessToolkit** [reply]
hi, yes you need to "just!" reset the MQTT state, after retrieving packages. 
![mqtt](https://github.com/PricelessToolkit/MailBoxGuard/assets/81021972/12b55a17-245d-436f-ac8b-915f8da803f3)

**Reply by steuerzentrale** [reply]
genius! many thanks for your great support, really appreciate that!
